Failure cluster [5014820e...]: Flake in TestTimeoutRequestHeaders and TestTimeoutWithLogging

### Failure cluster [5014820e8364ad0bace3](https://go.k8s.io/triage#5014820e8364ad0bace3)

Similar to https://github.com/kubernetes/kubernetes/issues/107991 and easily reproducible. The handler occasionally completes and sends a response before timing out.

##### Error text:
```
Failed
=== RUN   TestTimeoutRequestHeaders
    timeout_test.go:299: got res.StatusCde 200; expected 503
--- FAIL: TestTimeoutRequestHeaders (0.01s)

```
#### Recent failures:
[3/25/2023, 12:31:20 AM ci-kubernetes-unit-eks-canary](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-eks-canary/1639484447109681152)
[3/24/2023, 5:17:10 PM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/116899/pull-kubernetes-unit/1639375305921531904)
[3/21/2023, 9:05:34 AM ci-kubernetes-unit-1-24](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-24/1638163804372078592)
[3/18/2023, 7:39:50 PM ci-kubernetes-unit-1-23](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-23/1637236887804448768)
[3/15/2023, 9:59:17 PM ci-kubernetes-unit-eks-canary](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-eks-canary/1636184694812839936)


/kind failing-test
/kind flake

/sig api-machinery

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
@@ -272,6 +272,8 @@ func TestTimeoutRequestHeaders(t *testing.T) {
 		})
 	}
 
+	testDone := make(chan struct{})
+	defer close(testDone)
 	ts := httptest.NewServer(
 		withDeadline(
 			WithTimeoutForNonLongRunningRequests(
@@ -280,8 +282,13 @@ func TestTimeoutRequestHeaders(t *testing.T) {
 					cancel()
 					// mutate request Headers
 					// Authorization filter does it for example
-					for j := 0; j < 10000; j++ {
-						req.Header.Set("Test", "post")
+					for {
+						select {
+						case <-testDone:
+							return
+						default:
+							req.Header.Set("Test", "post")
+						}
 					}
 				}),
 				func(r *http.Request, requestInfo *request.RequestInfo) bool {
@@ -301,8 +308,8 @@ func TestTimeoutRequestHeaders(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if res.StatusCode != http.StatusGatewayTimeout {
-		t.Errorf("got res.StatusCde %d; expected %d", res.StatusCode, http.StatusServiceUnavailable)
+	if actual, expected := res.StatusCode, http.StatusGatewayTimeout; actual != expected {
+		t.Errorf("got status code %d; expected %d", actual, expected)
 	}
 	res.Body.Close()
 }
@@ -323,6 +330,8 @@ func TestTimeoutWithLogging(t *testing.T) {
 		})
 	}
 
+	testDone := make(chan struct{})
+	defer close(testDone)
 	ts := httptest.NewServer(
 		WithHTTPLogging(
 			withDeadline(
@@ -332,8 +341,13 @@ func TestTimeoutWithLogging(t *testing.T) {
 						cancel()
 						// mutate request Headers
 						// Authorization filter does it for example
-						for j := 0; j < 10000; j++ {
-							req.Header.Set("Test", "post")
+						for {
+							select {
+							case <-testDone:
+								return
+							default:
+								req.Header.Set("Test", "post")
+							}
 						}
 					}),
 					func(r *http.Request, requestInfo *request.RequestInfo) bool {
@@ -354,8 +368,8 @@ func TestTimeoutWithLogging(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	if res.StatusCode != http.StatusGatewayTimeout {
-		t.Errorf("got res.StatusCode %d; expected %d", res.StatusCode, http.StatusServiceUnavailable)
+	if actual, expected := res.StatusCode, http.StatusGatewayTimeout; actual != expected {
+		t.Errorf("got status code %d; expected %d", actual, expected)
 	}
 	res.Body.Close()
 }

